packages/coreutils/src/restorablepool.ts

```diff
@@ -0,0 +1,378 @@
+  /**
+   * Create a new restorable pool.
+   *
+   * @param options - The instantiation options for a restorable pool.
+   */
+  constructor(options: RestorablePool.IOptions) {
+    this.namespace = options.namespace;
+  }
+
+  /**
+   * A signal emitted when an object object is added.
+   *
+   * #### Notes
+   * This signal will only fire when an object is added to the pool.
+   * It will not fire if an object injected into the pool.
+   */
+  get added(): ISignal<this, T> {
+    return this._added;
+  }
+
+  /**
+   * A namespace for all tracked objects.
+   */
+  readonly namespace: string;
+
+  /**
+   * The current object.
+   *
+   * #### Notes
+   * The restorable pool does not set `current`. It is intended for client use.
+   *
+   * If `current` is set to an object that does not exist in the pool, it is a
+   * no-op.
+   */
+  get current(): T | null {
+    return this._current;
+  }
+  set current(obj: T) {
+    if (this._current === obj) {
+      return;
+    }
+    if (this._objects.has(obj)) {
+      this._current = obj;
+      this._currentChanged.emit(this._current);
+    }
+  }
+
+  /**
+   * A signal emitted when the current widget changes.
+   */
+  get currentChanged(): ISignal<this, T | null> {
+    return this._currentChanged;
+  }
+
+  /**
+   * A promise resolved when the restorable pool has been restored.
+   */
+  get restored(): Promise<void> {
+    return this._restored.promise;
+  }
+
+  /**
+   * The number of objects held by the pool.
+   */
+  get size(): number {
+    return this._objects.size;
+  }
+
+  /**
+   * A signal emitted when an object is updated.
+   */
+  get updated(): ISignal<this, T> {
+    return this._updated;
+  }
+
+  /**
+   * Add a new object to the pool.
+   *
+   * @param obj - The object object being added.
+   */
+  async add(obj: T): Promise<void> {
+    if (obj.isDisposed) {
+      const warning = 'A disposed object cannot be added.';
+      console.warn(warning, obj);
+      throw new Error(warning);
+    }
+
+    if (this._objects.has(obj)) {
+      const warning = 'This object already exists in the pool.';
+      console.warn(warning, obj);
+      throw new Error(warning);
+    }
+
+    this._objects.add(obj);
+    obj.disposed.connect(this._onInstanceDisposed, this);
+
+    if (Private.injectedProperty.get(obj)) {
+      return;
+    }
+
+    if (this._restore) {
+      const { connector } = this._restore;
+      const objName = this._restore.name(obj);
+
+      if (objName) {
+        const name = `${this.namespace}:${objName}`;
+        const data = this._restore.args(obj);
+
+        Private.nameProperty.set(obj, name);
+        await connector.save(name, { data });
+      }
+    }
+
+    // Emit the added signal.
+    this._added.emit(obj);
+  }
+
+  /**
+   * Test whether the pool is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the pool.
+   */
+  dispose(): void {
```

Comment:
Should this also clear the data connector? I'm not sure. Either way, I think it should be documented.

Reply:
I don't think it should clear the data connector. I'll add a note.